File: ip_jump_controller_qt/src/mockserver.cpp
#include "mockserver.h"
#include <QHostAddress>

MockServer::MockServer(MockServerState *state, QObject *parent)
    : QObject(parent), m_state(state)
{
}

MockServer::~MockServer()
{
    stop();
}

void MockServer::start()
{
    if (m_state->isRunning) return;

    m_server = new QTcpServer(this);
    QHostAddress addr(m_state->bindAddress);
    if (!m_server->listen(addr, m_state->port)) {
        LogEntry e;
        e.time = QDateTime::currentDateTime();
        e.direction = "out";
        e.summary = "Failed to start server: " + m_server->errorString();
        m_state->addLog(e);
        return;
    }

    m_state->isRunning = true;
    LogEntry e;
    e.time = QDateTime::currentDateTime();
    e.direction = "out";
    e.summary = "Server started on " + m_state->bindAddress + ":" + QString::number(m_state->port);
    m_state->addLog(e);

    connect(m_server, &QTcpServer::newConnection, this, &MockServer::onNewConnection);
}

void MockServer::stop()
{
    if (m_server) {
        m_server->close();
        m_server->deleteLater();
        m_server = nullptr;
    }
    m_buffers.clear();
    m_state->isRunning = false;

    LogEntry e;
    e.time = QDateTime::currentDateTime();
    e.direction = "out";
    e.summary = "Server stopped";
    m_state->addLog(e);
}

void MockServer::onNewConnection()
{
    while (m_server->hasPendingConnections()) {
        QTcpSocket *socket = m_server->nextPendingConnection();

        connect(socket, &QTcpSocket::readyRead, this, [this, socket]() {
            m_buffers[socket].append(socket->readAll());
            QByteArray &buf = m_buffers[socket];

            int headerEnd = buf.indexOf("\r\n\r\n");
            if (headerEnd < 0) return;

            // Parse Content-Length
            QByteArray headers = buf.left(headerEnd);
            int contentLength = 0;
            for (const QByteArray &line : headers.split('\n')) {
                if (line.trimmed().toLower().startsWith("content-length:")) {
                    int colonIdx = line.indexOf(':');
                    contentLength = line.mid(colonIdx + 1).trimmed().toInt();
                    break;
                }
            }

            int bodyStart = headerEnd + 4;
            if (buf.size() < bodyStart + contentLength) return;

            // Complete request
            QByteArray requestData = buf.left(bodyStart + contentLength);
            buf.remove(0, bodyStart + contentLength);
            handleRequest(socket, requestData);
        });

        connect(socket, &QTcpSocket::disconnected, this, [this, socket]() {
            m_buffers.remove(socket);
            socket->deleteLater();
        });
    }
}

void MockServer::handleRequest(QTcpSocket *socket, const QByteArray &data)
{
    int headerEnd = data.indexOf("\r\n\r\n");
    QByteArray headerBlock = data.left(headerEnd);
    QByteArray body = data.mid(headerEnd + 4);

    QList<QByteArray> headerLines = headerBlock.split('\n');
    QString firstLine = headerLines.first().trimmed();
    QStringList parts = firstLine.split(' ');
    QString method = parts.value(0);
    QString path = parts.value(1);

    m_authHeader.clear();
    for (const QByteArray &line : headerLines) {
        if (line.trimmed().toLower().startsWith("authorization:")) {
            int colonIdx = line.indexOf(':');
            m_authHeader = line.mid(colonIdx + 1).trimmed();
            break;
        }
    }

    QJsonObject reqData;
    if (!body.isEmpty()) {
        QJsonParseError err;
        QJsonDocument doc = QJsonDocument::fromJson(body, &err);
        if (err.error == QJsonParseError::NoError && doc.isObject())
            reqData = doc.object();
    }

    m_state->requestCount++;

    QJsonObject response;
    if (path == "/v1/auth")
        response = handleAuth(reqData);
    else if (path == "/v1/gettask")
        response = handleGetTask();
    else if (path == "/v1/getIpJump")
        response = handleGetIpJump();
    else if (path == "/v1/putIpJump")
        response = handlePutIpJump(reqData);
    else if (path == "/v1/uploadIp")
        response = handleUploadIp(reqData);
    else if (path == "/v1/getToken")
        response = handleGetToken();
    else if (path == "/v1/reportTaskCompletion")
        response = handleTaskCompletion(reqData);
    else if (path == "/v1/uploadproc" || path == "/v1/upload/suffix/exe")
        response = handleUploadProcess(reqData);
    else if (path == "/v1/getconf" || path == "/v1/getprotect")
        response = handleGetConfig();
    else if (path == "/v1/closetask")
        response = handleCloseTask(reqData);
    else
        response = QJsonObject{{"code", "000000"}, {"msg", "success"}, {"data", QJsonObject{}}};

    sendJson(socket, response);

    LogEntry entry;
    entry.time = QDateTime::currentDateTime();
    entry.direction = "in";
    entry.path = path;
    QString authTag = m_authHeader.isEmpty() ? "" : " [token]";
    entry.summary = method + " " + path + authTag +
                    "\n  >> " + summarizeRequest(path, reqData) +
                    "\n  << " + summarizeResponse(response);
    QJsonObject logData;
    logData["request"] = reqData;
    logData["response"] = response;
    entry.data = logData;
    m_state->addLog(entry);
}

void MockServer::sendJson(QTcpSocket *socket, const QJsonObject &resp)
{
    QByteArray body = QJsonDocument(resp).toJson(QJsonDocument::Compact);
    QByteArray response = "HTTP/1.1 200 OK\r\n"
                          "Content-Type: application/json\r\n"
                          "Content-Length: " + QByteArray::number(body.size()) + "\r\n"
                          "\r\n";
    response += body;
    socket->write(response);
    socket->flush();
    socket->disconnectFromHost();
}

// ==================== Route Handlers ====================

QJsonObject MockServer::handleAuth(const QJsonObject &req)
{
    m_state->agentUid = req.value("uid").toString();
    m_state->agentMacid = req.value("macid").toString();
    m_state->agentHostName = req.value("host_name").toString();
    QString ipList = req.value("ip").toString();
    if (!ipList.isEmpty())
        m_state->lastAgentIp = ipList.split(',').first().trimmed();

    QJsonObject resp;
    resp["code"] = "000000";
    resp["msg"] = "success";
    QJsonObject data;
    data["token"] = "mock-token-" + QString::number(QDateTime::currentMSecsSinceEpoch());
    resp["data"] = data;
    return resp;
}

QJsonObject MockServer::handleGetTask()
{
    QJsonArray taskList;
    if (m_state->ipJumpTaskPending)
        taskList.append(37);

    QJsonObject resp;
    resp["code"] = "000000";
    resp["msg"] = "success";
    QJsonObject data;
    data["tasklist"] = taskList;
    resp["data"] = data;
    return resp;
}

QJsonObject MockServer::handleGetIpJump()
{
    if (m_state->hasCurrentInstruction) {
        JumpInstruction inst = m_state->currentInstruction;
        m_state->currentInstruction = JumpInstruction();
        m_state->hasCurrentInstruction = false;
        m_state->ipJumpTaskPending = false;
        m_state->totalJumpsSent++;

        if (!m_state->instructionQueue.isEmpty()) {
            m_state->currentInstruction = m_state->instructionQueue.takeFirst();
            m_state->hasCurrentInstruction = true;
            m_state->ipJumpTaskPending = true;
        } else if (m_state->cycleStrategy.enabled) {
            QString nextTarget = inst.targetIp.contains('/')
                ? inst.targetIp.split('/').first()
                : inst.targetIp;
            JumpInstruction next = m_state->cycleStrategy.nextInstruction(nextTarget);
            if (!next.isEmpty()) {
                m_state->currentInstruction = next;
                m_state->hasCurrentInstruction = true;
                m_state->ipJumpTaskPending = true;
            }
        }

        LogEntry e;
        e.time = QDateTime::currentDateTime();
        e.direction = "out";
        e.path = "/v1/getIpJump";
        e.summary = "QUEUE: " + QString::number(m_state->instructionQueue.size()) + " remaining";
        m_state->addLog(e);

        QJsonObject resp;
        resp["code"] = "000000";
        resp["msg"] = "success";
        resp["data"] = inst.toJson();
        return resp;
    }

    if (m_state->cycleStrategy.enabled) {
        JumpInstruction inst = m_state->cycleStrategy.nextInstruction(m_state->lastAgentIp);
        if (!inst.isEmpty()) {
            m_state->totalJumpsSent++;
            JumpInstruction next = m_state->cycleStrategy.nextInstruction(inst.targetIp);
            if (!next.isEmpty()) {
                m_state->currentInstruction = next;
                m_state->hasCurrentInstruction = true;
                m_state->ipJumpTaskPending = true;
            }
            QJsonObject resp;
            resp["code"] = "000000";
            resp["msg"] = "success";
            resp["data"] = inst.toJson();
            return resp;
        }
    }

    QString reason = m_state->cycleStrategy.enabled
        ? "cycle enabled but ipPool has only " + QString::number(m_state->cycleStrategy.ipPool.size()) +
          " IPs (need >=2), lastAgentIp=" + m_state->lastAgentIp
        : "no instruction queued and cycle not enabled";

    LogEntry e;
    e.time = QDateTime::currentDateTime();
    e.direction = "out";
    e.path = "/v1/getIpJump";
    e.summary = "EMPTY: " + reason;
    m_state->addLog(e);

    QJsonObject emptyInst;
    emptyInst["source_ip"] = "";
    emptyInst["target_ip"] = "";
    emptyInst["gateway"] = "";
    emptyInst["active_time"] = 0;
    emptyInst["aging_time"] = 2;
    emptyInst["mode"] = 1;
    QJsonObject resp;
    resp["code"] = "000000";
    resp["msg"] = "success";
    resp["data"] = emptyInst;
    return resp;
}

QJsonObject MockServer::handlePutIpJump(const QJsonObject &req)
{
    int status = req.value("status").toInt();
    QString srcIp = req.value("source_ip").toString();
    QString tgtIp = req.value("target_ip").toString();
    QString agentIp = req.value("agent_ip").toString();
    QString reason = req.value("reason").toString();

    m_state->lastJumpStatus = (status == 1) ? "SUCCESS" : "FAILED";
    if (!agentIp.isEmpty())
        m_state->lastAgentIp = agentIp.split(',').first().trimmed();

    LogEntry e;
    e.time = QDateTime::currentDateTime();
    e.direction = "in";
    e.path = "/v1/putIpJump";
    e.summary = "Jump result: status=" + QString::number(status) +
                ", src=" + srcIp + ", tgt=" + tgtIp +
                ", agent=" + agentIp + ", reason=" + reason;
    e.data = req;
    m_state->addLog(e);

    return QJsonObject{{"code", "000000"}, {"msg", "success"}};
}

QJsonObject MockServer::handleUploadIp(const QJsonObject &req)
{
    QString ipList = req.value("ip").toString();
    if (!ipList.isEmpty()) {
        QStringList ips = ipList.split(',');
        if (!ips.isEmpty()) {
            m_state->lastAgentIp = ips.first().trimmed();
            LogEntry e;
            e.time = QDateTime::currentDateTime();
            e.direction = "in";
            e.path = "/v1/uploadIp";
            e.summary = "IP report: logical_primary=" + ips.first() + ", all=" + ipList;
            e.data = req;
            m_state->addLog(e);
        }
    }
    return QJsonObject{{"code", "000000"}, {"msg", "success"}};
}

QJsonObject MockServer::handleGetToken()
{
    QJsonObject resp;
    resp["code"] = "000000";
    resp["msg"] = "success";
    QJsonObject data;
    data["token"] = "mock-token-" + QString::number(QDateTime::currentMSecsSinceEpoch());
    resp["data"] = data;
    return resp;
}

QJsonObject MockServer::handleTaskCompletion(const QJsonObject &)
{
    return QJsonObject{{"code", "000000"}, {"msg", "success"}};
}

QJsonObject MockServer::handleUploadProcess(const QJsonObject &)
{
    return QJsonObject{{"code", "000000"}, {"msg", "success"}};
}

QJsonObject MockServer::handleGetConfig()
{
    return QJsonObject{{"code", "000000"}, {"msg", "success"}, {"data", QJsonObject{}}};
}

QJsonObject MockServer::handleCloseTask(const QJsonObject &)
{
    return QJsonObject{{"code", "000000"}, {"msg", "success"}};
}

// ==================== Summaries ====================

QString MockServer::summarizeRequest(const QString &path, const QJsonObject &req)
{
    if (path == "/v1/auth")
        return "AUTH uid=" + req.value("uid").toString("?") +
               ", host=" + req.value("host_name").toString("?") +
               ", ip=" + req.value("ip").toString("?");
    if (path == "/v1/gettask")
        return "GET-TASK (agent polling for task list)";
    if (path == "/v1/getIpJump")
        return "GET-IP-JUMP (agent fetching jump instruction)";
    if (path == "/v1/putIpJump")
        return "PUT-IP-JUMP status=" + QString::number(req.value("status").toInt()) +
               ", src=" + req.value("source_ip").toString() +
               ", tgt=" + req.value("target_ip").toString() +
               ", agent_ip=" + req.value("agent_ip").toString();
    if (path == "/v1/uploadIp")
        return "UPLOAD-IP: " + req.value("ip").toString();
    if (path == "/v1/reportTaskCompletion")
        return "TASK-COMPLETE";
    if (path == "/v1/closetask")
        return "CLOSE-TASK";
    return path + " (body keys=" +
           QStringList(req.keys()).join(',') + ")";
}

QString MockServer::summarizeResponse(const QJsonObject &resp)
{
    QString code = resp.value("code").toString();
    QJsonObject data = resp.value("data").toObject();
    if (data.contains("tasklist"))
        return "code=" + code + ", tasklist=" +
               QString(QJsonDocument(data["tasklist"].toArray()).toJson(QJsonDocument::Compact));
    if (data.contains("target_ip"))
        return "code=" + code + ", target=" + data["target_ip"].toString() +
               ", mode=" + QString::number(data["mode"].toInt()) +
               ", active=" + QString::number(data["active_time"].toInt()) + "s";
    if (data.contains("token")) {
        QString t = data["token"].toString();
        return "code=" + code + ", token=" +
               (t.length() > 20 ? t.left(20) + "..." : t);
    }
    return "code=" + code;
}

File: ip_jump_controller_egui/src/mockserver.rs
use std::io::Read;
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};

use tiny_http::{Request, Response};

// ==================== Data ====================

#[derive(Clone, serde::Serialize, serde::Deserialize, Default)]
pub struct LogEntry {
    pub time: String,
    pub direction: String,
    pub path: String,
    pub summary: String,
    pub data: Option<serde_json::Value>,
}

#[derive(Clone, serde::Serialize, serde::Deserialize, Default)]
pub struct JumpInstruction {
    pub source_ip: String,
    pub target_ip: String,
    pub gateway: String,
    pub mode: i32,
    pub active_time: i32,
    pub aging_time: i32,
    pub prefix: i32,
}

impl JumpInstruction {
    pub fn target_ip_cidr(&self) -> String {
        if self.target_ip.contains('/') {
            self.target_ip.clone()
        } else {
            format!("{}/{}", self.target_ip, self.prefix)
        }
    }

    pub fn source_ip_bare(&self) -> String {
        if self.source_ip.is_empty() {
            return String::new();
        }
        self.source_ip.split('/').next().unwrap_or("").to_string()
    }

    pub fn to_json(&self) -> serde_json::Value {
        serde_json::json!({
            "source_ip": self.source_ip_bare(),
            "target_ip": self.target_ip_cidr(),
            "gateway": self.gateway,
            "mode": self.mode,
            "active_time": self.active_time,
            "aging_time": self.aging_time,
        })
    }

    pub fn is_empty(&self) -> bool {
        self.target_ip.is_empty()
    }
}

#[derive(Clone, Default)]
pub struct CycleStrategy {
    pub enabled: bool,
    pub ip_pool: Vec<String>,
    pub gateway: String,
    pub mode: i32,
    pub active_time: i32,
    pub aging_time: i32,
    pub prefix: i32,
    pub current_index: usize,
}

impl CycleStrategy {
    pub fn next_instruction(&mut self, current_agent_ip: &str) -> JumpInstruction {
        if self.ip_pool.len() < 2 {
            return JumpInstruction::default();
        }
        let start_idx = self
            .ip_pool
            .iter()
            .position(|ip| ip == current_agent_ip)
            .unwrap_or(self.current_index);
        let next_idx = (start_idx + 1) % self.ip_pool.len();
        self.current_index = next_idx;
        JumpInstruction {
            target_ip: self.ip_pool[next_idx].clone(),
            gateway: self.gateway.clone(),
            mode: self.mode,
            active_time: self.active_time,
            aging_time: self.aging_time,
            prefix: self.prefix,
            ..Default::default()
        }
    }
}

// ==================== State ====================

pub struct InnerState {
    pub is_running: bool,
    pub port: u16,
    pub bind_address: String,
    pub agent_host_name: String,
    pub agent_uid: String,
    pub agent_macid: String,
    pub last_agent_ip: String,
    pub last_jump_status: String,
    pub request_count: u64,
    pub total_jumps_sent: u64,
    pub instruction_queue: Vec<JumpInstruction>,
    pub current_instruction: Option<JumpInstruction>,
    pub ip_jump_task_pending: bool,
    pub cycle_strategy: CycleStrategy,
    pub logs: Vec<LogEntry>,
}

impl Default for InnerState {
    fn default() -> Self {
        Self {
            is_running: false, port: 8080, bind_address: "0.0.0.0".into(),
            agent_host_name: String::new(), agent_uid: String::new(), agent_macid: String::new(),
            last_agent_ip: String::new(), last_jump_status: String::new(),
            request_count: 0, total_jumps_sent: 0,
            instruction_queue: vec![], current_instruction: None, ip_jump_task_pending: false,
            cycle_strategy: CycleStrategy::default(), logs: vec![],
        }
    }
}

pub type SharedState = Arc<Mutex<InnerState>>;

// ==================== Helpers ====================

pub fn log_add(state: &SharedState, entry: LogEntry) {
    let mut s = state.lock().unwrap();
    s.logs.insert(0, entry);
    if s.logs.len() > 500 { s.logs.truncate(500); }
}

fn now_ms() -> u128 {
    SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis()
}

fn summarize_req(path: &str, req: &serde_json::Value) -> String {
    let s = |k: &str| req.get(k).and_then(|v| v.as_str()).unwrap_or("");
    match path {
        "/v1/auth" => format!("AUTH uid={}, host={}, ip={}", s("uid"), s("host_name"), s("ip")),
        "/v1/gettask" => "GET-TASK".into(),
        "/v1/getIpJump" => "GET-IP-JUMP".into(),
        "/v1/putIpJump" => format!("PUT-IP-JUMP status={}, src={}, tgt={}, agent_ip={}",
            req.get("status").and_then(|v| v.as_i64()).unwrap_or(0), s("source_ip"), s("target_ip"), s("agent_ip")),
        "/v1/uploadIp" => format!("UPLOAD-IP: {}", s("ip")),
        _ => format!("{path}"),
    }
}

fn summarize_resp(resp: &serde_json::Value) -> String {
    let code = resp["code"].as_str().unwrap_or("");
    let data = &resp["data"];
    if data.get("tasklist").is_some() { return format!("code={code}, tasklist={}", data["tasklist"]); }
    if data.get("target_ip").is_some() {
        return format!("code={code}, target={}, mode={}, active={}s",
            data["target_ip"].as_str().unwrap_or(""), data["mode"].as_i64().unwrap_or(0),
            data["active_time"].as_i64().unwrap_or(0));
    }
    if let Some(t) = data.get("token").and_then(|v| v.as_str()) {
        return format!("code={code}, token={}", if t.len() > 20 { &t[..20] } else { t });
    }
    format!("code={code}")
}

// ==================== Request Router ====================

pub fn route_request(state: &SharedState, mut request: Request) {
    let path = request.url().to_string();
    let method = request.method().clone();
    let method_str = method.as_str().to_string();

    let mut body = String::new();
    let _ = request.as_reader().read_to_string(&mut body);
    let req_data: serde_json::Value = serde_json::from_str(&body).unwrap_or(serde_json::Value::Null);

    let auth = request.headers().iter()
        .find(|h| h.field.as_str().as_str().to_lowercase() == "authorization")
        .map(|h| h.value.as_str().to_string()).unwrap_or_default();

    // Collect pending log entries
    let mut pending_logs: Vec<LogEntry> = Vec::new();
    let now = || chrono::Local::now().format("%H:%M:%S").to_string();
    let resp: serde_json::Value;
    {
        let mut s = state.lock().unwrap();
        s.request_count += 1;

        match path.as_str() {
            "/v1/auth" => {
                s.agent_uid = req_data["uid"].as_str().unwrap_or("").to_string();
                s.agent_macid = req_data["macid"].as_str().unwrap_or("").to_string();
                s.agent_host_name = req_data["host_name"].as_str().unwrap_or("").to_string();
                let ips = req_data["ip"].as_str().unwrap_or("");
                if !ips.is_empty() { s.last_agent_ip = ips.split(',').next().unwrap_or("").to_string(); }
                resp = serde_json::json!({"code":"000000","msg":"success",
                    "data":{"token":format!("mock-token-{}",now_ms())}});
            }
            "/v1/gettask" => {
                let mut tl: Vec<i32> = vec![];
                if s.ip_jump_task_pending { tl.push(37); }
                resp = serde_json::json!({"code":"000000","msg":"success","data":{"tasklist":tl}});
            }
            "/v1/getIpJump" => {
                // Extract full state needed for this handler
                let inst = s.current_instruction.take();
                let has_queue = !s.instruction_queue.is_empty();
                let cyc_enabled = s.cycle_strategy.enabled;
                let last_agent_ip = s.last_agent_ip.clone();

                if let Some(inst) = inst {
                    s.total_jumps_sent += 1;
                    let nxt = inst.target_ip_cidr().split('/').next().unwrap_or("").to_string();
                    if has_queue {
                        s.current_instruction = Some(s.instruction_queue.remove(0));
                        s.ip_jump_task_pending = true;
                    } else if cyc_enabled {
                        let next = s.cycle_strategy.next_instruction(&nxt);
                        if !next.is_empty() {
                            s.current_instruction = Some(next);
                            s.ip_jump_task_pending = true;
                        } else { s.ip_jump_task_pending = false; }
                    } else { s.ip_jump_task_pending = false; }
                    resp = serde_json::json!({"code":"000000","msg":"success","data":inst.to_json()});
                    pending_logs.push(LogEntry {
                        time: now(), direction: "out".into(), path: "/v1/getIpJump".into(),
                        summary: format!("QUEUE: {} remaining", s.instruction_queue.len()),
                        data: None,
                    });
                } else if cyc_enabled {
                    let mut inst = s.cycle_strategy.next_instruction(&last_agent_ip);
                    if !inst.is_empty() {
                        s.total_jumps_sent += 1;
                        let next = s.cycle_strategy.next_instruction(&inst.target_ip);
                        if !next.is_empty() {
                            s.current_instruction = Some(next);
                            s.ip_jump_task_pending = true;
                        }
                    }
                    // fudge inst back (moved out)
                    inst = if !inst.is_empty() {
                        JumpInstruction { target_ip: last_agent_ip.clone(), ..Default::default() }
                    } else { JumpInstruction::default() };
                    let (inst_data, empty, reason) = if !inst.is_empty() {
                        (inst.to_json(), false, String::new())
                    } else {
                        let reason = format!("cycle enabled but ipPool has only {} IPs (need >=2), lastAgentIp={last_agent_ip}",
                            s.cycle_strategy.ip_pool.len());
                        (serde_json::json!({"source_ip":"","target_ip":"","gateway":"","active_time":0,"aging_time":2,"mode":1}), true, reason)
                    };
                    resp = serde_json::json!({"code":"000000","msg":"success","data":inst_data});
                    if empty {
                        pending_logs.push(LogEntry { time: now(), direction: "out".into(),
                            path: "/v1/getIpJump".into(), summary: format!("EMPTY: {reason}"), data: None });
                    }
                } else {
                    let reason = "no instruction queued and cycle not enabled";
                    resp = serde_json::json!({"code":"000000","msg":"success",
                        "data":{"source_ip":"","target_ip":"","gateway":"","active_time":0,"aging_time":2,"mode":1}});
                    pending_logs.push(LogEntry { time: now(), direction: "out".into(),
                        path: "/v1/getIpJump".into(), summary: format!("EMPTY: {reason}"), data: None });
                }
            }
            "/v1/putIpJump" => {
                let status = req_data["status"].as_i64().unwrap_or(0);
                let agent_ip = req_data["agent_ip"].as_str().unwrap_or("");
                s.last_jump_status = if status == 1 { "SUCCESS".into() } else { "FAILED".into() };
                if !agent_ip.is_empty() { s.last_agent_ip = agent_ip.split(',').next().unwrap_or("").to_string(); }
                resp = serde_json::json!({"code":"000000","msg":"success"});
                pending_logs.push(LogEntry {
                    time: now(), direction: "in".into(), path: "/v1/putIpJump".into(),
                    summary: format!("Jump result: status={status}, src={}, tgt={}, agent={}, reason={}",
                        req_data["source_ip"].as_str().unwrap_or(""),
                        req_data["target_ip"].as_str().unwrap_or(""),
                        req_data["agent_ip"].as_str().unwrap_or(""),
                        req_data["reason"].as_str().unwrap_or("")),
                    data: Some(req_data.clone()),
                });
            }
            "/v1/uploadIp" => {
                let ips = req_data["ip"].as_str().unwrap_or("");
                if !ips.is_empty() {
                    s.last_agent_ip = ips.split(',').next().unwrap_or("").to_string();
                    let la = s.last_agent_ip.clone();
                    pending_logs.push(LogEntry {
                        time: now(), direction: "in".into(), path: "/v1/uploadIp".into(),
                        summary: format!("IP report: logical_primary={la}, all={ips}"),
                        data: Some(req_data.clone()),
                    });
                }
                resp = serde_json::json!({"code":"000000","msg":"success"});
            }
            "/v1/getToken" => {
                resp = serde_json::json!({"code":"000000","msg":"success",
                    "data":{"token":format!("mock-token-{}",now_ms())}});
            }
            "/v1/reportTaskCompletion" | "/v1/uploadproc" | "/v1/upload/suffix/exe"
            | "/v1/getconf" | "/v1/getprotect" | "/v1/closetask" => {
                resp = serde_json::json!({"code":"000000","msg":"success"});
            }
            _ => {
                resp = serde_json::json!({"code":"000000","msg":"success","data":{}});
            }
        }
    }

    // Add main request log
    let auth_tag = if auth.is_empty() { "" } else { " [token]" };
    let main_log = LogEntry {
        time: now(), direction: "in".into(), path: path.clone(),
        summary: format!("{} {path}{auth_tag}\n  >> {}\n  << {}",
            method_str, summarize_req(&path, &req_data), summarize_resp(&resp)),
        data: Some(serde_json::json!({"request": req_data, "response": resp.clone()})),
    };
    {
        let mut s = state.lock().unwrap();
        s.logs.insert(0, main_log);
        for l in pending_logs { s.logs.insert(0, l); }
        if s.logs.len() > 500 { s.logs.truncate(500); }
    }

    let resp_body = serde_json::to_string(&resp).unwrap_or_default();
    let response = Response::from_string(resp_body).with_status_code(200);
    let _ = request.respond(response);
}
